esrally/track/params.py

```diff
@@ -368,6 +369,86 @@ def params(self):
+            try:
+                template = params["template"]
+                self.template_definitions.append(template)
+            except KeyError:
+                raise exceptions.InvalidSyntax(f"Please set the property 'template' for the {params.get('operation-type')} operation.")
+
+    def params(self):
+        return {
+            "templates": self.template_definitions,
+            "only-if-exists": self.only_if_exists,
+            "request-params": self.request_params
+        }
+
+
+class CreateComposableTemplateParamSource(ParamSource):
+    def __init__(self, track, params, templates=None, **kwargs):
+        templates = track.composable_templates if templates is None else templates
+        super().__init__(track, params, **kwargs)
+        self.request_params = params.get("request-params", {})
+        self.template_definitions = []
+        if "template" in params and "body" in params:
+            self.template_definitions.append((params["template"], params["body"]))
+        elif templates:
+            filter_template = params.get("template")
+            settings = params.get("settings")
+            for template in templates:
+                if not filter_template or template.name == filter_template:
+                    body = template.content
+                    if body and "template" in body:
+                        body = CreateComposableTemplateParamSource._create_or_merge(template.content,
+                                                                                    ["template", "settings"], settings)
+                        self.template_definitions.append((template.name, body))
+        else:
+            raise exceptions.InvalidSyntax("Please set the properties 'template' and 'body' for the "
+                                           f"{params.get('operation-type')} operation or declare composable and/or component "
+                                           "templates in the track")
+
+    @staticmethod
+    def _create_or_merge(content, path, new_content):
+        original_content = content
+        if new_content:
+            for sub_path in path:
+                if sub_path not in content:
+                    content[sub_path] = {}
+                content = content[sub_path]
+            CreateComposableTemplateParamSource.__merge(content, new_content)
+        return original_content
+
+    @staticmethod
+    def __merge(dct, merge_dct):
+        for k in merge_dct.keys():
+            if (k in dct and isinstance(dct[k], dict)
+                    and isinstance(merge_dct[k], collections.Mapping)):
+                CreateComposableTemplateParamSource.__merge(dct[k], merge_dct[k])
+            else:
+                dct[k] = merge_dct[k]
+
+    def params(self):
+        return {
+            "templates": self.template_definitions,
+            "request-params": self.request_params
+        }
+
+
+class CreateComponentTemplateParamSource(CreateComposableTemplateParamSource):
```

Comment:
How about we use an abstract base class and derive two sub classes from it? I think such a class hierarchy would be more consistent.

Reply:
I agree this is more consistent with the actual concept